### ISSUE TITLE
CORE-1 — Typed outputs + declarative observable specification

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1555]
**Depends on:** INF-1 (#1550) · **Blocks:** ARCH-3 (#1562), TRN-2 (#1576), CFG-1 (#1574)

**Context:** Today model outputs are unvalidated dicts. **Eleven** legacy model classes return `Dict[str, Optional[torch.Tensor]]` from `forward` — `MACE`, `ScaleShiftMACE`, `AtomicDipolesMACE`, `AtomicDielectricMACE`, `EnergyDipolesMACE` (`mace/modules/models.py:289,471,778,1065,1345`) and `MACELES`, `PolarMACE`, `MagneticMACE`, `MagneticScaleShiftMACE`, `MagneticSCFMACE`, `TimeReversalSymmetrizedMACE` (`mace/modules/extensions.py:291,966,1707,1759,2029,2278`) — the last of which is not CLI-reachable and is the one a hand-written enumeration drops — and between them they emit **43 distinct string keys**, from `energy`/`forces`/`stress` through `magforces`, `BEC`, `latent_{charges,dipoles,alphas,kappas,quads}`, `external_field`, `fukui_functions`, `polarizability_sh` and the SCF trio `scf_energy_history`/`scf_steps`/`equilibrated_magmom`. Only **40** of the 43 are keys of a returned dict literal: the SCF trio is assigned onto the wrapped model's output after the fact (`extensions.py:2107-2112`), so a count read off return literals stops at 40 and silently loses the SCF surface. And the model forward is only the innermost of **three** layers users can see — the calculator and the eval CLI each add names of their own (constraint 7). Every consumer keeps its own key lists by hand: the ASE calculator maintains hardcoded `graph_level_keys` / `atom_level_keys` sets to slice padding (`mace/calculators/mace.py:463-491`, applied at `:496-499`) which classify **22 of those 43 keys** — the other 21 fall through the `else` branch at `:500-501` and are returned with their padding rows intact. Which observables a model computes is decided by a string ladder: `mace/cli/run_train.py:588-625` branches on the model-class name (`AtomicDipolesMACE`, `AtomicDielectricMACE`, `EnergyDipolesMACE`, `PolarMACE` + loss name) and sets six `compute_*` booleans per branch, while the forward signatures carry ten more (`compute_force`, `compute_virials`, `compute_stress`, `compute_displacement`, `compute_hessian`, `compute_edge_forces`, `compute_atomic_stresses`, `compute_magforces`, `compute_bec`, `compute_dielectric_derivatives`, plus `training` and `lammps_mliap`) — so adding an output property means editing multiple files across the codebase. This ticket builds the two abstractions that replace all of that — a typed, framework-generic output structure and a declarative observable spec. They are the reforge's most consequential new abstractions and feed the whole downstream chain: spec → automatic head creation (ARCH-3 (#1562)) → automatic loss term (TRN-2 (#1576)) → training config (CFG-1 (#1574)). Keep them minimal — no abstraction "for the future".

**Interface & constraints:**

1. **`MACEOutput(Generic[TensorT])`** in `mace_core.outputs`. `TensorT = TypeVar(...)` so the same class carries `torch.Tensor` in mace-torch and `jax.Array` in mace-jax — `mace_core` itself imports neither (a minimal `Protocol` for the tensor type is allowed if needed). Optional core fields: `total_energy`, `node_energies`, `forces`, `stress`, `virials`, `dipole`. Everything else goes through the escape hatch `extras: dict[str, TensorT]` — custom data never breaks the strict typing of core fields. The 43-key legacy surface is what makes `extras` load-bearing rather than decorative: six core fields plus a typed hatch is the deliberate split, and the electrostatic, magnetic and dielectric families live in the hatch with their `per_atom`/`per_graph` classification carried by their `ObservableSpec`, not by a second hardcoded key set. Models never return raw dicts. Implement as a plain dataclass unless validation is actually needed — justify the dataclass-vs-Pydantic choice in the PR. This `Generic[TensorT]` trick is load-bearing beyond this ticket: the kernel Protocol reuses exactly the same pattern to stay framework-free, so it must genuinely work with no framework installed.

2. **`ObservableSpec`** in `mace_core.observables`, a Pydantic model with fields: `name`, `irreps` (string spec of the spherical-tensor shape in the canonical irreps grammar — e.g. `"1o"` for a per-atom vector, `"128x0e+128x1o+128x2e"` for a sum of multiplicities; this ticket stores and grammar-validates the string, it does not implement irreps algebra), `per_atom` (bool), `units` (project convention: eV, Å), `normalization`, `default_loss_weight` (seeds the automatic loss term in TRN-2 (#1576)). **`normalization`** is a required string enum — one of `"none"`, `"std"`, `"rms"` — naming how the target quantity is scaled before it reaches a head and a loss term; it is the per-observable successor of the legacy scaling registry (`std_scaling`, `rms_forces_scaling`, `rms_dipoles_scaling`, the three entries of `scaling_classes` at `mace/modules/__init__.py:91-95`). This ticket only stores and validates the value: the head applies it in the output layer and the loss applies the matching term, both reading this one field, so no second scaling mechanism appears anywhere. Any atomic or total spherical-tensor property declared this way must become trainable with zero new code downstream.

3. **Derivative auto-naming, over declared inputs — not just position and cell.** For a declared quantity `<q>` differentiated with respect to a declared input `<x>`, the derivative is auto-named `d_<q>_d_<x>`, with `pos` and `cell` as the two inputs every model has. Three named special cases, all data rather than code branches scattered in consumers:

| quantity | input | canonical name | sign |
|---|---|---|---|
| `energy` | positions | `forces` | −1 |
| `energy` | cell (strain) | `stress` | +1 |
| `energy` | `magmom` | `magforces` | −1 |

   The third is not an extrapolation: develop computes `magforces` as `−∂E/∂magmom` in the same autograd call as forces (`mace/modules/utils.py:262-299`, `compute_forces_magforces`, returning `-forces, -mag_forces`), it is a first-class loss term (`mace/modules/loss.py:453-458,483-488`) and the SCF model drives its fixed point with it (`extensions.py:2077`). A grammar that only knows `d_<q>_d_pos` and `d_<q>_d_cell` cannot express the magnetic family at all, and MAG-1 (#1584) would then have to bypass the spec — which is the failure this abstraction exists to prevent. Declaring an input feature (`magmom` today; `elec_temp`, `total_charge`, `total_spin` and the `embedding_specs` features tomorrow) and asking for its derivative must require zero new code.

4. **Defaults file.** Ship `defaults/observables.yaml` declaring energy plus its position and cell derivatives — the canonical example every other observable copies — with a loader and tests.

5. **Error contract, and the padding classification.** Validation errors must name the offending observable and state the accepted grammar (agent-friendly: name the key/value and the fix). The downstream hard error — declaring a property that is absent from the data — is enforced at head creation (ARCH-3 (#1562)); this ticket's spec must carry enough metadata (`name`, `per_atom`, `irreps`) to make that check and the automatic padding/unpadding of graph-level vs node-level fields possible. That consumer is DATA-3 (#1571)/DEP-1 (#1583), and the bar it has to clear is explicit: the derived classification must cover **every** key the model emits, so the 21-key fall-through in `mace/calculators/mace.py:500-501` cannot recur. Unclassifiable output is an error at spec time, never a silent pass-through of padded rows.

6. **Purity is tested, not assumed.** Tests exercise `MACEOutput` with numpy arrays (proves framework-agnosticism), and one test asserts `"torch" not in sys.modules` after importing `mace_core`.

7. **The user-observable output surface is three layers, and the completeness gate is defined over all three.** "v1 has the same functionality as develop" is a claim about the names a user can read out, not about the names one function happens to return — a key that reaches the user only through the calculator or only through `mace_eval_configs` is functionality just the same, and a gate keyed on the model forward alone cannot see it. The set is therefore defined as the union of:

   - **(a) model forward** — every string key of the dict returned by `forward` on the eleven model classes in `mace/modules/{models,extensions}.py`, counting both dict-literal keys and keys assigned onto the returned object before it is returned. **43 keys.** Nine of the eleven contribute: `MagneticMACE.forward` (`extensions.py:1697-1708`) is abstract and raises `NotImplementedError`, and `TimeReversalSymmetrizedMACE.forward` (`extensions.py:2278`) returns the wrapped model's keys unchanged, so neither adds a name.
   - **(b) ASE calculator `results`** — every key written into `self.results` by `MACECalculator.calculate` (`mace/calculators/mace.py:727-890`) and `MagneticMACECalculator.calculate` (`:1444-1450`). **31 keys**, of which **15 exist at this layer only**: `free_energy` and `energies` (the aliases and the E0-inclusive per-atom energy, `:801-804`), `stresses` (the Voigt per-atom rename of the model's `atomic_stresses`, `:735-739`), `LES_alphas` / `LES_kappas` (`:832-839`), `bec` (the lower-cased mean of the model's `BEC`, `:840`), `MACE_magmoms` (`:1467`), and the eight committee keys `{energy,forces,stress,dipole}_{comm,var}` (`:771,:781`). Read the committee set off the *code*, not off `implemented_properties`: the loop emits both suffixes for all four members of `results_store_ensemble`, while `implemented_properties` advertises only four of the eight (`:278-293`) — so `forces_var`, `stress_comm` and `dipole_comm` are produced and never declared.
   - **(c) `mace_eval_configs` output** — every name written into `atoms.arrays` or `atoms.info`, after the `--info_prefix` prefix (default `"MACE_"`, `mace/cli/eval_configs.py:103-107`), at `eval_configs.py:423-472`. **13 names** over 14 write sites (`descriptors` is written to `info` or to `arrays` depending on `--descriptor_aggregation_method`), of which **3 exist at this layer only** and each is a *rename* of a model key, which is exactly why they are easy to lose: `BO_contributions` (model `contributions`), `descriptors` (model `node_feats`, after invariant extraction and layer truncation), `node_energies` (model `node_energy`). Only `energy`, `forces` and `stress` also appear as calculator `results` keys — the other **ten reach the user through this CLI alone**: the three renames plus `BEC`, `latent_charges`, `latent_dipoles`, `latent_kappas`, `latent_alphas`, `latent_quads` and `magforces`, for which the calculator offers at most a differently-named subset (`bec`, `LES_alphas`, `LES_kappas`) and usually nothing.

   **Union: 61 distinct names** — 43 + 15 + 3. That number, not 43, is what "the output surface survived the rewrite" is measured against.

   **Ownership, one ticket per layer.** This ticket owns (a): the `ObservableSpec` completeness test enumerates the 43 mechanically and maps each to a spec or a DROP row. **DEP-1 (#1583)** owns the energy-family part of (b) and **DEP-1a (#1634)** the dipole/dielectric/polar/LES/magnetic part; each must account for its layer's calculator-only names against the spec rather than re-growing a hand-kept list. **CLI-1 (#1579)** owns (c) for `mace eval`, with the P0-5 eval contract as its behavioural floor. The three enumerations are the same mechanical extraction run at three sites, so each ticket re-derives its layer from the source rather than trusting a number copied from here.

**Task:**
1. **`mace_core.outputs`:** implement `MACEOutput(Generic[TensorT])` per constraint 1 (dataclass unless validation is needed — justify the choice in the PR); no torch/jax imports; numpy-based tests plus the `sys.modules` purity assertion.
2. **`mace_core.observables`:** implement the Pydantic `ObservableSpec` per constraints 2–3, including the `d_<q>_d_<x>` derivative auto-naming with the three named special cases and their signs; ship `defaults/observables.yaml` declaring energy + position/cell derivatives, with loader and tests; validation errors name the observable and the accepted grammar.
3. Add the completeness test: the 43 legacy output keys enumerated from `mace/modules/{models,extensions}.py` — enumerated mechanically, covering keys assigned onto the output as well as dict literals, so the SCF trio is in the set — each map to exactly one `ObservableSpec` shape (`per_atom` true/false) or to an explicit DROP row with a reason; no key is unaccounted for. Ship the enumeration as a reusable helper, since DEP-1 (#1583)/DEP-1a (#1634) and CLI-1 (#1579) run the same extraction over layers (b) and (c) of constraint 7. Keys DROPped here rather than specified: the SCF diagnostics `scf_energy_history` / `scf_steps` (run telemetry, not observables — they belong in the SCF stage's log, and MAG-1 (#1584) carries that decision), and `equilibrated_magmom`, which is an *input* the SCF stage converged, exposed through the stage result rather than as a model observable.

**Out of scope:** head creation (ARCH-3 (#1562)), loss construction (TRN-2 (#1576)), config machinery (CORE-2 (#1556)), the SCF fixed-point stage itself (MAG-1 (#1584)/TRN-2 (#1576)).

**Acceptance criteria:**
- [ ] Declaring a new rank-2 per-atom observable purely in YAML parses to a valid spec with derived `d_<q>_d_pos` / `d_<q>_d_cell` names, with zero code changes.
- [ ] Declaring a new *input* feature and asking for `d_energy_d_<feature>` parses with zero code changes; the `magmom` → `magforces` (−1) case is covered by a test alongside forces and stress.
- [ ] The forces/stress special case (naming and the −1 sign on the position derivative of energy) is covered by tests.
- [ ] All 43 legacy model-forward output keys are either specified or carry an explicit DROP row with a reason (completeness test), the 43 re-derived from the source by the test rather than hard-coded, so a new key added to a legacy forward fails it.
- [ ] The three-layer surface of constraint 7 is written down where the downstream tickets read it: the 61-name union, its 43/15/3 split, and which ticket owns each layer.
- [ ] `mace-core` imports without torch installed; a test asserts `"torch" not in sys.modules` after importing it; `MACEOutput` tests run on numpy arrays.
- [ ] A malformed spec produces a validation error that names the observable and the accepted grammar.

**Verify:**
```bash
python -m pytest packages/mace-core/tests/test_outputs.py packages/mace-core/tests/test_observables.py -v
ruff check packages/mace-core && ty check packages/mace-core
```
**Review focus:** cross-review — each half (outputs, observables) reviewed by someone who didn't write it. Watch for scope creep: these two abstractions must stay minimal, and the derivative grammar generalising past `pos`/`cell` is the one place where the extra generality is paid for by an existing model family, not by speculation.

---
